subsys/fs/nffs_fs.c

```diff
@@ -0,0 +1,540 @@
+	}
+
+	/* We need to assume all bytes were written */
+	return size;
+}
+
+int fs_seek(fs_file_t *zfp, off_t offset, int whence)
+{
+	u32_t len;
+	u32_t pos;
+	int rc;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	switch (whence) {
+	case FS_SEEK_SET:
+		pos = offset;
+		break;
+	case FS_SEEK_CUR:
+		pos = zfp->fp->nf_offset + offset;
+		break;
+	case FS_SEEK_END:
+		rc = nffs_inode_data_len(zfp->fp->nf_inode_entry, &len);
+		if (rc) {
+			k_mutex_unlock(&nffs_lock);
+			return -EINVAL;
+		}
+		pos = len + offset;
+		break;
+	default:
+		k_mutex_unlock(&nffs_lock);
+		return -EINVAL;
+	}
+
+	rc = nffs_file_seek(zfp->fp, pos);
+
+	k_mutex_unlock(&nffs_lock);
+
+	return translate_error(rc);
+}
+
+off_t fs_tell(fs_file_t *zfp)
+{
+	u32_t offset;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	if (!zfp->fp) {
+		return -EIO;
+		k_mutex_unlock(&nffs_lock);
+	}
+
+	offset = zfp->fp->nf_offset;
+
+	k_mutex_unlock(&nffs_lock);
+
+	return offset;
+}
+
+int fs_truncate(fs_file_t *zfp, off_t length)
+{
+	/*
+	 * FIXME:
+	 * There is no API in NFFS to truncate opened file. For now we return
+	 * ENOTSUP, but this should be revisited if truncation is implemented
+	 * in NFFS at some point.
+	 */
+
+	return -ENOTSUP;
+}
+
+int fs_sync(fs_file_t *zfp)
+{
+	/*
+	 * Files are written to flash immediately so we do not need to support
+	 * sync call - just return success.
+	 */
+
+	return 0;
+}
+
+int fs_mkdir(const char *path)
+{
+	int rc;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	if (!nffs_misc_ready()) {
+		k_mutex_unlock(&nffs_lock);
+		return -ENODEV;
+	}
+
+	rc = nffs_path_new_dir(path, NULL);
+
+	k_mutex_unlock(&nffs_lock);
+
+	return translate_error(rc);
+}
+
+int fs_opendir(fs_dir_t *zdp, const char *path)
+{
+	int rc;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	zdp->dp = NULL;
+
+	if (!nffs_misc_ready()) {
+		k_mutex_unlock(&nffs_lock);
+		return -ENODEV;
+	}
+
+	rc = nffs_dir_open(path, &zdp->dp);
+
+	k_mutex_unlock(&nffs_lock);
+
+	return translate_error(rc);
+}
+
+int fs_readdir(fs_dir_t *zdp, struct fs_dirent *entry)
+{
+	struct nffs_dirent *dirent;
+	int rc;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	rc = nffs_dir_read(zdp->dp, &dirent);
+	switch (rc) {
+	case 0:
+		rc = inode_to_dirent(dirent->nde_inode_entry, entry);
+		break;
+	case FS_ENOENT:
+		entry->name[0] = 0;
+		rc = 0;
+		break;
+	default:
+		goto done;
+	}
+
+done:
+	k_mutex_unlock(&nffs_lock);
+
+	return translate_error(rc);
+}
+
+int fs_closedir(fs_dir_t *zdp)
+{
+	int rc;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	rc = nffs_dir_close(zdp->dp);
+	if (!rc) {
+		zdp->dp = NULL;
+	}
+
+	k_mutex_unlock(&nffs_lock);
+
+	return translate_error(rc);
+}
+
+int fs_stat(const char *path, struct fs_dirent *entry)
+{
+	struct nffs_path_parser parser;
+	struct nffs_inode_entry *parent;
+	struct nffs_inode_entry *inode;
+	int rc;
+
+	k_mutex_lock(&nffs_lock, K_FOREVER);
+
+	nffs_path_parser_new(&parser, path);
+
+	rc = nffs_path_find(&parser, &inode, &parent);
+	if (rc) {
+		goto done;
```

Comment:
`goto` dose not fit to any style in this file and it is too trivial case for use it, rearange thi to if (!rc) ...